food-truck-api/src/main/java/food/truck/api/endpoint/TruckEndpoint.java

```diff
@@ -226,4 +233,58 @@ public void deleteTruckRouteLocations(@AuthenticationPrincipal User user, @Reque
     public Route getTodaysRoute(@PathVariable long truckId) {
         return truckService.getActiveRoute(truckId, LocalDateTime.now().getDayOfWeek());
     }
+
+    @Secured({"ROLE_USER"})
+    @PostMapping("/truck/{truckId}/subscribe")
+    public Optional<SubscriptionView> subscribe(@AuthenticationPrincipal User u, @PathVariable long truckId) {
+        var t = truckService.findTruckById(truckId);
+        if (t.isEmpty()) {
+            return Optional.empty();
+        }
+        var truck = t.get();
+
+        // Prevent an owner from subscribing to their own truck.
+        if (!u.getId().equals(truck.getUserId())) {
+            var sub = new Subscription();
+            sub.setTruck(truck);
+            sub.setUser(u);
+
+            return Optional.of(subscriptionService.saveSubscription(sub)).map(SubscriptionView::of);
+        } else {
+            return Optional.empty();
+        }
+    }
+
+    @Secured({"ROLE_USER"})
+    @GetMapping("/truck/{truckId}/subscription")
+    public Optional<SubscriptionView> subscribedToTruck(@AuthenticationPrincipal User u, @PathVariable long truckId) {
+        var subs = subscriptionService.findSubsByUser(u);
+        var iter = subs.stream()
+            .filter(sub -> sub.getTruck().getId().equals(truckId))
+            .map(SubscriptionView::of)
+            .iterator();
+        if (iter.hasNext()) {
+            return Optional.of(iter.next());
+        }
+        return Optional.empty();
+    }
+
+    @Secured({"ROLE_USER"})
+    @DeleteMapping("/truck/{truckId}/unsubscribe")
+    public void unsubscribe(@AuthenticationPrincipal User u, @PathVariable long truckId) {
+        var t = truckService.findTruckById(truckId);
+        if (t.isEmpty()) {
+            return;
+        }
+        var truck = t.get();
+        var subs = subscriptionService.findSubsByUser(u);
+        var filteredSubs = subs
+            .stream()
+            .filter(sub -> sub.getTruck().equals(truck))
```

Comment:
I'd personally just compare the `truckIds` here, since comparing Longs will be substantially cheaper than comparing trucks, and this operation could potentially be called several times in a production environment for one unsubscribe. However, not really a big deal for us here.

Reply:
I agree. Looking at this in hindsight that would be the better approach.